src/Execution/DataLoader/BatchLoaderRegistry.php

```diff
@@ -21,10 +21,15 @@ abstract class BatchLoaderRegistry
      *
      * @throws \Exception
      */
-    public static function instance(string $loaderClass, array $pathToField, array $constructorArgs = []): object
+    public static function instance(string $loaderClass, string $relationName, array $pathToField, array $relationScopes = [], array $fieldArguments, array $constructorArgs = []): object
```

Comment:
Everything about relations should probably go into `RelationBatchLoader`. That keeps this class generically useful.

Reply:
You are right, but i'm not quite sure how this should work.  Currently the BatchLoaderRegistry generates the instance key before even creating a RelationBatchLoader instance.  And only long after the instance is created with it's key in the registry do we register the relation info inside the RelationBatchLoader.
`
        /** @var \Nuwave\Lighthouse\Execution\DataLoader\RelationBatchLoader $relationBatchLoader */
        $relationBatchLoader = BatchLoaderRegistry::instance(RelationBatchLoader::class, $uniquePath);

        if (! $relationBatchLoader->hasRelationLoader()) {
            $relationBatchLoader->registerRelationLoader($this->relationLoader($resolveInfo), $relationName);
        }
`

I have a couple ideas though:

1) BatchLoaderRegistry@Instance accepts $constructorArgs for building the BatchLoader.  Instead of registering the relation data after, maybe we could:
          a) add a constructor to RelationBatchLoader and use this instead of the registerRelationLoader method later
          b) create a batchKeySuffix method directly on RelationBatchLoader that allows it to partially handle part of the instanceKey generation itself.
          c) Pass RelationBatchLoader data through BatchLoaderRegistry@instance $constructorArgs
          d) Have BatchLoaderRegistry@Instance generate the RelationBatchLoader instance with $constructorArgs
          e) HaveBatchLoaderRegistry@instance generate the key normally based on the path (with the field).
          f) Check if a batchKeySuffix method exists on the RelationBatchLoader and if it does remove the last part of the path (the field) and replace it with the batchKeySuffix.

2) A more simple solution we could just have withRelationDirective add all the relation data to the path array before it is passed into BatchLoaderRegistry@instance.  The current code is already doing something like this with the relation name (i think maybe to allow the use of multiple @with directives on the same field).

Let me know what you think of either of those solutions.

Reply:
Solution 2 sent me on the track I am now on and lead to an even cleaner solution. Thanks.